transcom/mymove
# B 19924 tio too service level policy int

## [B-19924](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A948236)

## Summary

Enforces the uniqueness of either TIO or TOO role on a user at the service level. Only one or the other of these roles may exist, never both at the same time.

### Smaller details
Added detailed validation error responses via the API instead of just internal server errors. Includes new validation enforcement and refactoring of "users_roles". Soft deletion of user roles now take place prior to attempting to add, this is so that validation rules on "add user roles" don't consider roles that are about to be deleted. Due to [B-19923](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A948235) and [B-19922](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A948234) being completed first, this is only GUI testable on the main version of this branch. The steps to replicate will utilize postman.

## Verification Steps for Reviewers

These are to be checked by a reviewer.


### How to test

1. Open postman -> ghc api -> open -> transportation-offices -> get -> adjust URL search query to grab a transportation office
2. Take the returned transportation office ID and keep it somewhere for now, we’ll need this later when you paste the template json bodies
3. Log into adminlocal
4. Grab the latest X-Csrf-Token from your network tab
5. Open postman -> admin api -> office-users -> create an office user
6. Add the X-Csrf-Token to your headers
7. Add the “Create office user” body
8. Adjust transportationOfficeId based on what you grabbed earlier
9. Attempt to create with both TOO and TIO roles, you should receive a detailed verr response
10. Remove one of the “TIO” or “TOO” roles from the body
11. Post again, this time it should succeed
12. Keep this request tab open as it holds your office user id as well as the payload you used to create the user
13. Navigate to a new endpoint in a new postman tab, navigate admin api -> office users ->  update a user
14. Update the URL and put your office user ID in (Or use the headers if you want). The office user id is the “id” value in the response payload, not the “userId” value.
15. Update the body with exactly what you used for the creation
16. Update your X-Csrf-Token in the headers (Since this is a new tab you won’t have it populated), you should be able to take it from your previous request. If not, grab the latest from the browser again
17. But this time put back in the deleted role so you violate the TIO/TOO policy
18. Send the patch request and you should receive a detailed verr
19. Adjust your payload and remove the role that you created the office user with, it should now only hold the opposite transportation officer role that the user does not have
20. Submit and it should update the user to now have only that new role you provided
21. To verify, navigate to a new endpoint: admin api -> office-users -> officeUserId -> GET -> Paste your office user id into the URL
22. Submit get request, and you should see the role successfully
23. Repeat for the office API endpoint of office users

Create office user body
```
{
  "email": "user@example.com",
  "firstName": "string",
  "middleInitials": "L.",
  "lastName": "string",
  "telephone": "212-555-5555",
  "transportationOfficeId": "9804dbaa-71b0-4170-8ae2-fcba1356fb7f",
  "roles": [
    {
      "name": "Transportation Ordering Officer",
      "roleType": "transportation_ordering_officer"
    },
    {
      "name": "Transportation Invoicing Officer",
      "roleType": "transportation_invoicing_officer"
    }
  ]
}
```
